Source/Particles/Collision/BinaryCollision/NuclearFusionFunc.H

```diff
@@ -0,0 +1,204 @@
+#include <AMReX_Vector.H>
+
+/**
+ * \brief This functor does binary nuclear fusions on a single cell.
+ *  Particles of the two reacting species are paired with each other and for each pair we compute
+ *  if a fusion event occurs. If so, we fill a mask (input parameter p_mask) with true so that
+ *  product particles corresponding to a given pair can be effectively created in the particle
+ *  creation functor.
+ *  This functor also reads and contains the fusion multiplier.
+ */
+class NuclearFusionFunc{
+    // Define shortcuts for frequently-used type names
+    using ParticleType = WarpXParticleContainer::ParticleType;
+    using ParticleBins = amrex::DenseBins<ParticleType>;
+    using index_type = ParticleBins::index_type;
+    using SoaData_type = WarpXParticleContainer::ParticleTileType::ParticleTileDataType;
+
+public:
+    /**
+     * \brief Default constructor of the NuclearFusionFunc class.
+     */
+    NuclearFusionFunc () = default;
+
+    /**
+     * \brief Constructor of the NuclearFusionFunc class
+     *
+     * @param[in] collision_name the name of the collision
+     * @param[in] mypc pointer to the MultiParticleContainer
+     * @param[in] isSameSpecies whether the two colliding species are the same
+     */
+    NuclearFusionFunc (const std::string collision_name, MultiParticleContainer* const mypc,
+                       const bool isSameSpecies) : m_isSameSpecies(isSameSpecies)
+    {
+        using namespace amrex::literals;
+
+        m_fusion_type = BinaryCollisionUtils::get_nuclear_fusion_type(collision_name, mypc);
+
+        amrex::ParmParse pp_collision_name(collision_name);
+        amrex::Vector<std::string> product_species_name;
+        pp_collision_name.getarr("product_species", product_species_name);
+
+        if (m_fusion_type == NuclearFusionType::ProtonBoron)
+        {
+            AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
+                product_species_name.size() == 1,
+                "ERROR: Proton-boron must contain exactly one product species");
+            auto& product_species = mypc->GetParticleContainerFromName(product_species_name[0]);
+            AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
+                product_species.AmIA<PhysicalSpecies::helium>(),
+                "ERROR: Product species of proton-boron fusion must be of type helium");
+        }
+
+        // default fusion multiplier
+        m_fusion_multiplier = 1.0_rt;
+        queryWithParser(pp_collision_name, "fusion_multiplier", m_fusion_multiplier);
+        // default fusion probability threshold
+        m_probability_threshold = 0.02_rt;
+        queryWithParser(pp_collision_name, "fusion_probability_threshold",
+                                            m_probability_threshold);
+        // default fusion probability target_value
+        m_probability_threshold = 0.002_rt;
+        queryWithParser(pp_collision_name, "fusion_probability_target_value",
+                                            m_probability_target_value);
```

Comment:
These initial values should keep the error coming from the use of a fusion multiplier below the percent level. I plan on adding a more detailed explanation somewhere in the WarpX documentation in a follow-up PR.